### ISSUE TITLE
Documentation for custom cache implementations

### ISSUE THREAD

**Issue by migueloller** [apollographql/apollo-client#2484]
Hi there! 👋 

I was wondering if there was a plan to document the API for custom cache implementations.

As a huge fan of Apollo and happy user of 1.0, I can't wait to migrate to 2.0 and take advantage of all the awesome new features. Unfortunately, I won't be able to do that until there's a cache implementation with Redux.

Where would be the best place to start learning about the contract needed for the Apollo cache so that a Redux version can be implemented? Also, is the 1.0 implementation close enough so that a Redux implementation of the cache can be extracted from there?

**Comment by jbaxleyiii** [comment]
@migueloller we absolutely are planning on documenting it 🎉 @kamilkisiela has even started work on an rx-js variant of the InMemoryCache that we should be able to adapt as a guide for redux integration.

I am curious, why is the cache being in redux a blocker for you?

**Comment by btroia** [comment]
@jbaxleyiii I'm in the same boat. I need query persistence for certain queries when a user quits/restarts my app. Previously I was using redux-persist to include Apollo data, but with 2.0 all queries will refetch once each time my app is restarted. Or am I missing something? Thanks.

**Comment by migueloller** [comment]
@jbaxleyiii, awesome! Great to hear that's on the works.

As to why the Redux cache is a blocker, there are a couple of reasons.
- We use tools in the Redux ecosystem to manage various things related to state/side effects, including but not limited to:
  - Tracking errors in a central location and sending them to an error reporting service like Sentry
  - Adding offline functionality using tools like redux-persist like @btroia mentioned
- Using Redux's benefits of time-travel and a better development experience by having state changes and side effects logged by middleware.

There are various things we use that I'm not mentioning here. Also, one could argue that we could move away from these features and use the new programming patterns that Apollo 2.0 allows. Ultimately, though, it would be optimal to adopt these patterns incrementally. Having a cache that uses Redux internally would allow a smooth transition from 1.0 to 2.0 and then we can slowly transition to new patterns as we see fit.

Hopefully these are reasons enough to justify the implementation of a Redux cache. 😅 

tl;dr: Having a Redux cache would allow to take advantage of the mature Redux ecosystem and tools while smoothly transitioning into new patterns from Apollo 2.0 as needed.

**Comment by simontegg** [comment]
I'm also using redux-persist. 
I'm not sure if you consider this bad practice but I also sometimes find it useful to use [reselect](https://github.com/reactjs/reselect) for querying data that depends in apollo's redux cache, rather than write a whole query.

**Comment by lancygoyal** [comment]
GraphQL + Apollo + Redux + React = 😊

**Comment by entropitor** [comment]
@jbaxleyiii Another reason why redux can be useful is so we can use reselect as to only compute the denormalized state if the relevant normalized state has changed

**Comment by jbmeslin** [comment]
@jbaxleyiii in my case i catch the APOLLO_QUERY_RESULT action  in a redux middleware in which i can inspect the extensions of my query. I got some anonymized fields then i call an other graphQL endpoint and finally i dispatch a second time this action with my resolved fields. If there is a solution to do that in V2 (or a better way in V1 ;) ) i agree to say redux is no more needed.

**Comment by jbaxleyiii** [comment]
@jbmeslin so you get a result, fire another query, then write to the cache with that data? That can be done using an apollo-link now!

**Comment by jbmeslin** [comment]
Ok i have to spend some times to understand how works apollo-link and reproduce my use case. 

 I call => uri1/graphqlX  ->  {dataX: {},  extensionsX, }  in dataX i have some anonymized field listed in extensionsX then i call an other endPoint to resolved these fields  => uir2/graphQLY-> {dataY}   with this result replace in dataX the anonymized fields with the resolved fields in dataY. 

The second query use another URI and the result structure of dataY is not the same as dataX.

**Comment by Amareis** [comment]
Hey, anybody? Where are docs for custom cache implementations? I find some interfaces at https://github.com/apollographql/apollo-client/blob/master/packages/apollo-cache/src/cache.ts but it very bad docs :)

**Comment by hwillson** [comment]
@Amareis The custom cache implementation docs don't exist (yet). This is still something we'd love to get in place though. If anyone is interested in helping out, this issue is now `help-wanted` 🙏 🙇. Thanks!

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests / discussions and bugs, and to help clean up the feature request / discussion backlog, Apollo Client feature requests / discussions are now being managed under the https://github.com/apollographql/apollo-feature-requests repository. 

This feature request / discussion will be closed here, but anyone interested in migrating this issue to the new repository (to make sure it stays active), can click [here](https://github.com/apollographql/apollo-feature-requests/issues/new?title=Documentation%20for%20custom%20cache%20implementations&body=Migrated%20from:%20apollographql/apollo-client%232484) to start the migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!

**Comment by fbartho** [comment]
Migrated to: https://github.com/apollographql/apollo-feature-requests/issues/47#issue-357360701
